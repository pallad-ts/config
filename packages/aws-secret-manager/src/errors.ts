import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

export const errorsDomain = new Domain();

const code = formatCodeFactory("E_CONF_SECRET_MANAGER_%c");
export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
	CANNOT_EXTRACT_PROPERTY_FROM_NON_OBJECT: ErrorDescriptor.useDefaultMessage(
		code(1),
		"Cannot extract property from non-object"
	),
	MISCONFIGURED_DATALOADER: ErrorDescriptor.useMessageFormatter(
		code(2),
		(batchSize: number) => `Misconfigured dataloader. Missing required "maxBatchSize: ${batchSize}" option`
	),
});
