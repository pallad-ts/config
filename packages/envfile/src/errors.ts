import {Domain, generators} from 'alpha-errors';
import {ConfigError} from "@pallad/config";

export const ERRORS = new Domain({
    errorClass: ConfigError,
    codeGenerator: generators.formatCode('E_CONF_ENVFILE_%d')
}).createErrors(create => {
    return {
        ENV_FILE_DOES_NOT_EXIST: create('Env file %s does not exist but it is required'),
    };
});
