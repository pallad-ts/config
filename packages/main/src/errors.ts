import {Domain, generators} from 'alpha-errors';
import {ConfigError} from "./ConfigError";

export const ERRORS = new Domain({
    errorClass: ConfigError,
    codeGenerator: generators.formatCode('E_CONF_%d')
}).createErrors(create => {
    return {
        CONFIG_LOADING_FAILED: create('Config loading failed: \s%s'),
        PROVIDER_VALUE_NOT_AVAILABLE: create('Value not available: %s'),
        SYNC_LOADING_NOT_AVAILABLE: create('Cannot load config synchronously since one of your config value providers is asynchronous'),
        PICK_PROVIDER_UNREGISTERED_TYPE: create('Unregistered type "%s". Registered values: %s'),
        PICK_PROVIDER_TYPE_ALREADY_EXISTS: create('Type "%s" already registered'),
        CANNOT_CONVERT_TO_INT: create(`Value "%s" cannot be converted to int`),
        CANNOT_CONVERT_TO_NUMBER: create(`Value "%s" cannot be converted to number`),
        CANNOT_CONVERT_TO_BOOL: create(`Value "%s" cannot be converted to bool`),
        TYPE_URL_INVALID_PROTOCOL: create('Protocol "%s" is not allowed. Allowed: %s'),
        INVALID_PORT_NUMBER: create('Port "%s" is invalid. Allowed ranges: %s')
    };
});
