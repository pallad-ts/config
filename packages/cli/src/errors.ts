import { ConfigError } from "@pallad/config";
import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

const errorsDomain = new Domain();
const code = formatCodeFactory("E_CONF_CLI_%c");

export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
    INVALID_CONFIG_VALUE_AT_PATH: ErrorDescriptor.useMessageFormatter(
        code(1),
        (parentConfigValue: unknown, propertyPath: string[]) =>
            `Cannot get config at property "${propertyPath.join(",")}" from config since parent value is of type: ${typeof parentConfigValue}`,
        ConfigError
    ),
});
