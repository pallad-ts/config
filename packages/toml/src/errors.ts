import { ConfigError } from "@pallad/config";
import { Domain, ErrorDescriptor, formatCodeFactory } from "@pallad/errors";

const errorsDomain = new Domain();
const code = formatCodeFactory("E_CONF_TOML_%c");

export const ERRORS = errorsDomain.addErrorsDescriptorsMap({
    ENV_FILE_DOES_NOT_EXIST: ErrorDescriptor.useMessageFormatter(
        code(1),
        (filePath: string) => `TOML file "${filePath}" does not exist but it is required`,
        ConfigError
    ),
});
