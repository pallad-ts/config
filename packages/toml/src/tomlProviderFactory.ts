import { all as mergeAll, Options as DeepMergeOptions } from "deepmerge";
import { existsSync, readFileSync } from "fs";
import { load as loadToml } from "js-toml";
import { resolve } from "path";

import { FilePathEntry, runOnlyOnceWrapper } from "@pallad/config-common";

import { TomlProvider } from "./TomlProvider";
import { ERRORS } from "./errors";

export function tomlProviderFactory(options: tomlProviderFactory.Options) {
    function loadConfig() {
        const cwd = options?.cwd ?? process.cwd();

        const files = Array.isArray(options.files) ? options.files : [options.files];
        const configList = [] as Array<Record<string, any>>;

        for (const filePathEntry of files) {
            const pathEntry = FilePathEntry.normalize(filePathEntry);

            const filePath = resolve(cwd, pathEntry.path);

            if (!existsSync(filePath)) {
                if (pathEntry.required) {
                    throw ERRORS.ENV_FILE_DOES_NOT_EXIST.create(pathEntry.path);
                }
                continue;
            }

            configList.push(loadToml(readFileSync(filePath, "utf-8")));
        }

        return mergeAll(configList, options.deepMerge) as Record<string, unknown>;
    }

    const load = runOnlyOnceWrapper(loadConfig);
    const factory = ((propertyPath: string) => {
        return new TomlProvider(propertyPath, loadConfig);
    }) as unknown as tomlProviderFactory.ProviderFactory;

    factory.load = load;
    return factory;
}

export namespace tomlProviderFactory {
    export interface ProviderFactory {
        (propertyPath: string): TomlProvider;

        load(): Record<string, unknown>;
    }

    export interface Options {
        files: FilePathEntry | FilePathEntry[];
        /**
         * Working directory. By default `process.cwd()`
         */
        cwd?: string;
        deepMerge?: DeepMergeOptions;
    }
}
