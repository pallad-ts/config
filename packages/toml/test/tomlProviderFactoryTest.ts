import { ERRORS } from "@src/errors";
import { tomlProviderFactory } from "@src/tomlProviderFactory";
import { realpathSync } from "node:fs";
import * as path from "node:path";

import "@pallad/errors-dev";

describe("tomlProviderFactory", () => {
    const FIXTURES_DIR = realpathSync(path.join(__dirname, "fixtures"));

    it("main factory produces provider", () => {
        const factory = tomlProviderFactory({
            files: [{ path: path.join(FIXTURES_DIR, "base.toml"), required: true }],
        });

        const provider = factory("foo");
        expect(provider.getValue().unwrap()).toBe("bar");
    });

    describe("loading", () => {
        it("fails if required file does not exist", () => {
            const factory = tomlProviderFactory({
                files: [{ path: path.join(FIXTURES_DIR, "non-existing.toml"), required: true }],
            });

            expect(() => {
                factory.load();
            }).toThrowErrorWithCode(ERRORS.ENV_FILE_DOES_NOT_EXIST);
        });

        it("basic loading", () => {
            const factory = tomlProviderFactory({
                files: [{ path: path.join(FIXTURES_DIR, "base.toml"), required: true }],
            });

            expect(factory.load()).toMatchSnapshot();
        });

        it("ignore non required and non existing files", () => {
            const factory = tomlProviderFactory({
                files: [{ path: path.join(FIXTURES_DIR, "non-existing.toml"), required: false }],
            });
            expect(factory.load()).toEqual({});
        });

        it("merges multiple files", () => {
            const factoryBase = tomlProviderFactory({
                files: [{ path: path.join(FIXTURES_DIR, "base.toml"), required: true }],
            });

            const factory = tomlProviderFactory({
                files: [
                    { path: path.join(FIXTURES_DIR, "base.toml"), required: true },
                    { path: path.join(FIXTURES_DIR, "section.toml"), required: true },
                ],
            });

            expect(factoryBase.load()).toHaveProperty("section.value", 1);
            expect(factory.load()).toHaveProperty("section.value", 2);

            expect(factory.load()).toHaveProperty("jwt", [
                { key: "key", value: "value" },
                { key: "key2", value: "value2" },
                { key: "key3", value: "value3" },
            ]);
        });

        it("customizing deep merge", () => {
            const factory = tomlProviderFactory({
                files: [
                    { path: path.join(FIXTURES_DIR, "base.toml"), required: true },
                    { path: path.join(FIXTURES_DIR, "section.toml"), required: true },
                ],
                deepMerge: {
                    customMerge: () => {
                        // always return original value
                        return a => a;
                    },
                },
            });

            expect(factory.load()).toHaveProperty("section.value", 1);
        });
    });
});
