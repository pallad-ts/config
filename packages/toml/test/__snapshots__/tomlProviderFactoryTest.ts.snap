// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tomlProviderFactory loading basic loading 1`] = `
{
  "foo": "bar",
  "jwt": [
    {
      "key": "key",
      "value": "value",
    },
    {
      "key": "key2",
      "value": "value2",
    },
  ],
  "section": {
    "value": 1,
  },
  "test": 1,
}
`;
